File: 2022/12/part1.py
heightmap = []
startpos = None
endpos = None

with open("input.txt") as inp:
    for line in inp:
        row = []
        for char in line.strip():
            if char.islower():
                row.append(ord(char) - ord('a'))
            elif char == 'S':
                startpos = (len(heightmap), len(row))
                row.append(0)
            elif char == 'E':
                endpos = (len(heightmap), len(row))
                row.append(25)
        heightmap.append(row)


rows = len(heightmap)
cols = len(heightmap[0])

unvisited = set((row, col) for row in range(rows) for col in range(cols))
current = startpos
moves = ((0, 1), (1, 0), (0, -1), (-1, 0))
INF = 1000000

distances = {}
for row in range(rows):
    for col in range(cols):
        distances[(row, col)] = INF
distances[startpos] = 0

while endpos in unvisited:
    row, col = current
    for move in moves:
        nrow, ncol = (row + move[0], col + move[1])
        if nrow < 0 or nrow >= rows or ncol < 0 or ncol >= cols:
            continue
        if heightmap[nrow][ncol] - heightmap[row][col] > 1:
            continue
        neighbour = (nrow, ncol)
        if neighbour not in unvisited:
            continue
        if distances[current] + 1 < distances[neighbour]:
            distances[neighbour] = distances[current] + 1

    unvisited.remove(current)

    for row in range(rows):
        for col in range(cols):
            if (row, col) == current:
                print("X", end="")
            elif (row, col) not in unvisited:
                print(".", end="")
            elif distances[(row, col)] == INF:
                print("#", end="")
            else:
                print("X", end="")
        print("")
    print("")

    mindist = INF
    for pos in unvisited:
        if distances[pos] < mindist:
            mindist = distances[pos]
            current = pos

print(distances[endpos])

File: 2025/07/part1.c
#include <stdio.h>

// #include "sample_input.c"
#include "input.c"

char map[ROWS][COLS] = {0};

int step(int row) {
  int split_count = 0;
  for (int col = 0; col < COLS; col++) {
    if (map[row][col] == 'S') {
      map[row][col] = '|';
      continue;
    }

    if (map[row - 1][col] == '|') {
      if (map[row][col] == '.') {
        map[row][col] = '|';
      } else if (map[row][col] == '^') {
        split_count++;
        if (col > 0)
          map[row][col - 1] = '|';
        if (col < COLS - 1)
          map[row][col + 1] = '|';
      }
    }
  }
  return split_count;
}

int main(void) {
  for (int row = 0; row < ROWS; row++)
    for (int col = 0; col < COLS; col++)
      map[row][col] = input_strings[row][col];

  int total = 0;
  for (int row = 0; row < ROWS; row++) {
    total += step(row);
  }
  printf("%d\n", total);

  return 0;
}
